internal/provider/kubernetes/controller.go

```diff
@@ -769,13 +769,37 @@ func (r *gatewayAPIReconciler) findReferenceGrant(ctx context.Context, from, to
 	}
 
 	for _, refGrant := range refGrants {
-		if refGrant.Namespace == to.namespace {
-			for _, src := range refGrant.Spec.From {
-				if src.Kind == gwapiv1a2.Kind(from.kind) && string(src.Namespace) == from.namespace {
-					return &refGrant, nil
-				}
+		if refGrant.Namespace != to.namespace {
+			continue
+		}
+
+		var fromAllowed bool
+		for _, refGrantFrom := range refGrant.Spec.From {
+			if string(refGrantFrom.Kind) == from.kind &&
+				string(refGrantFrom.Namespace) == from.namespace {
+				fromAllowed = true
 			}
 		}
+
+		if !fromAllowed {
+			continue
+		}
+
+		var toAllowed bool
+		for _, refGrantTo := range refGrant.Spec.To {
+			if string(refGrantTo.Kind) == to.kind &&
+				(refGrantTo.Name == nil ||
+					*refGrantTo.Name == "" ||
```

Comment:
let's keep this three conditions in one line

Reply:
changed!